### PR TITLE
:zap: shrink state size

### LINKED ISSUE #133: state size
Make the serialized state shrinker

### PR DESCRIPTION
**Pull request by fabienjuif**
I wanted to try to limit the state (serialized state) size.
So I try to only use the `data` field which is an object.

And then use `Object.keys` and `Object.values` to retrieve `array` and `keys` field, into selectors.
I know this is less performant, but I guess this is a good tradeoff for server side rendering (state is tinier in the payload when we want to send the state to the client after the first render -on server-)

But but but, `Object.values()` doesn't sort the values to the order defined into the object.
So I stop there for now.

I let the PR open so the idea still exists.

(FYI @bpetetot @EmrysMyrddin)

fixes #133

### CODE REVIEW

File: src/__snapshots__/selectors.spec.js.snap [hunk 1]
```diff
@@ -239,51 +249,20 @@ exports[`selectors should retrieve initialized value 1`] = `true`;
 
 exports[`selectors should retrieve keys 1`] = `
 Array [
+  "0",
```
Comment (fabienjuif) [inline src/__snapshots__/selectors.spec.js.snap]: Here 0 should not be the first item.

Comment (bpetetot) [reply src/__snapshots__/selectors.spec.js.snap]: It may be overkill , but the `Map` object in ES6 seems to guarantee the key/value order

Comment (fabienjuif) [reply src/__snapshots__/selectors.spec.js.snap]: _notes_
 - Map.set(key, value)
Something to try:
```
state {
  data: Array.from(map.entries()),
  initialized: <bool>,
}
```
  
+ selectors + memoisation
`Array.from(new Map(state).values())`
`Array.from(new Map(state).keys())`

**Comment by fabienjuif** [comment]
- first version with Map
 - I have to check snapshots a little closer
 - I have to add memoize on some selectors

**Comment by fabienjuif** [comment]
One selector is a breaking change, this is `getState` because it retrieves `state` as it is (without any transformation), and before it was:
```js
{
  data: {
    '1': { some: 'value', id: '1' },
  },
  keys: ['1'],
  array: [
    { some: 'value', id: '1' },
  ],
  initialized: true,
}
```

Now this is:
```js
{
  data: [
    [1, { some: 'value', id: '1' }],
  ]
  initialized: true,
}
```

So if someone use this selectors instead of high level one, this gona break

**Comment by fabienjuif** [comment]
Perf are ... shitty

```
// BEFORE (master)
results (ms): {
      "writes": {
        "total": 3993,
        "init": 3,
        "set": 59,
        "add": 1027,
        "update": 1186,
        "remove": 1030,
        "addOrUpdate": 688
      },
      "reads": {
        "total": 22610,
        "getState": 14,
        "getKeys": 15,
        "getAsArray": 20,
        "getLength": 17,
        "isInitialized": 11,
        "get": 18,
        "getByIds": 22437,
        "getById": 40,
        "hasKey": 38
      },
      "all": 26603
    }

// RERUN_SELECTORS / 100
results (ms): {
      "writes": {
        "total": 1363,
        "init": 3,
        "set": 33,
        "add": 342,
        "update": 371,
        "remove": 273,
        "addOrUpdate": 341
      },
      "reads": {
        "total": 12034,
        "getState": 1,
        "getKeys": 2583,
        "getAsArray": 2471,
        "getLength": 0,
        "isInitialized": 0,
        "get": 2323,
        "getByIds": 1589,
        "getById": 1535,
        "hasKey": 1531
      },
      "all": 13397
    }

// RERUN_SELECTORS / 100 with memoization
results (ms): {
      "writes": {
        "total": 1370,
        "init": 3,
        "set": 108,
        "add": 270,
        "update": 374,
        "remove": 266,
        "addOrUpdate": 349
      },
      "reads": {
        "total": 6925,
        "getState": 1,
        "getKeys": 3,
        "getAsArray": 3,
        "getLength": 1,
        "isInitialized": 0,
        "get": 2291,
        "getByIds": 1571,
        "getById": 1530,
        "hasKey": 1525
      },
      "all": 8295
    }
```

After memoization reading perf is 50 x slower

**Comment by fabienjuif** [comment]
https://gist.github.com/fabienjuif/f836ae2a499629168b20ffc370fbb83b

**Comment by fabienjuif** [comment]
With better memoization (I was dumb)
```
 results (ms): {
      "writes": {
        "total": 1290,
        "init": 4,
        "set": 34,
        "add": 281,
        "update": 373,
        "remove": 260,
        "addOrUpdate": 338
      },
      "reads": {
        "total": 2717,
        "getState": 9,
        "getKeys": 12,
        "getAsArray": 16,
        "getLength": 27,
        "isInitialized": 19,
        "get": 14,
        "getByIds": 2539,
        "getById": 42,
        "hasKey": 39
      },
      "all": 4007
    }
```

So we run faster with this version memoized since `getByIds` is faster

File: src/types/keyValue/keyValue.selectors.js [hunk 2]
```diff
@@ -1 +1,72 @@
-export { getState, getKeys, getAsArray, getLength, isInitialized, get, getBy, hasKey } from '../../selectors'
+const getFromPath = (data, path) => path.split('.').reduce(
+  (curr, sub) => curr && curr[sub],
+  data,
+)
+
+const memoize = (callback) => {
+  const memory = {
+    args: undefined,
```
Comment (bpetetot) [inline src/types/keyValue/keyValue.selectors.js]: you named it `data` when you are using it bellow

**Comment by fabienjuif** [comment]
With last commit.

What is your though @bpetetot @frinyvonnick @EmrysMyrddin  ?

```json
    results (ms): {
      "writes": {
        "total": 428,
        "init": "0.03ms / op [1000]",
        "sets": "0.11ms / op [1000]",
        "add": "0.22ms / op [1]",
        "adds": "0.48ms / op [1000]",
        "update": "0.46ms / op [1]",
        "updates": "1.32ms / op [1000]",
        "remove": "0.29ms / op [1]",
        "removes": "0.23ms / op [1000]",
        "addOrUpdate": "0.15ms / op [1]",
        "addOrUpdates": "0.99ms / op [1000]"
      },
      "reads": {
        "total": 1990,
        "getState": "0.12ms / op [1]",
        "getKeys": "0.16ms / op [1000]",
        "getAsArray": "0.15ms / op [1]",
        "getLength": "0.14ms / op [1]",
        "isInitialized": "0.13ms / op [1]",
        "get": "0.55ms / op [1]",
        "getById": "0.73ms / op [1]",
        "getByIds": "17.34ms / op [1000]",
        "hasKey": "0.58ms / op [1]"
      },
      "all": 2418
    }
```

**Comment by fabienjuif** [comment]
With last commit.

What is your though @bpetetot @frinyvonnick @EmrysMyrddin  ?

```json
    results (ms): {
      "writes": {
        "total": 428,
        "init": "0.03ms / op [1000]",
        "sets": "0.11ms / op [1000]",
        "add": "0.22ms / op [1]",
        "adds": "0.48ms / op [1000]",
        "update": "0.46ms / op [1]",
        "updates": "1.32ms / op [1000]",
        "remove": "0.29ms / op [1]",
        "removes": "0.23ms / op [1000]",
        "addOrUpdate": "0.15ms / op [1]",
        "addOrUpdates": "0.99ms / op [1000]"
      },
      "reads": {
        "total": 1990,
        "getState": "0.12ms / op [1]",
        "getKeys": "0.16ms / op [1000]",
        "getAsArray": "0.15ms / op [1]",
        "getLength": "0.14ms / op [1]",
        "isInitialized": "0.13ms / op [1]",
        "get": "0.55ms / op [1]",
        "getById": "0.73ms / op [1]",
        "getByIds": "17.34ms / op [1000]",
        "hasKey": "0.58ms / op [1]"
      },
      "all": 2418
    }
```

This is master:

```json
   results (ms): {
      "writes": {
        "total": 419,
        "init": "0.02ms / op [1000]",
        "sets": "0.06ms / op [1000]",
        "add": "0.16ms / op [1]",
        "adds": "0.33ms / op [1000]",
        "update": "0.35ms / op [1]",
        "updates": "1.37ms / op [1000]",
        "remove": "0.35ms / op [1]",
        "removes": "0.26ms / op [1000]",
        "addOrUpdate": "0.15ms / op [1]",
        "addOrUpdates": "1.12ms / op [1000]"
      },
      "reads": {
        "total": 2380,
        "getState": "0.00013ms / op [1]",
        "getKeys": "0.00015ms / op [1000]",
        "getAsArray": "0.00014ms / op [1]",
        "getLength": "0.00014ms / op [1]",
        "isInitialized": "0.00013ms / op [1]",
        "get": "0.00055ms / op [1]",
        "getById": "0.00075ms / op [1]",
        "getByIds": "0.02123ms / op [1000]",
        "hasKey": "0.00058ms / op [1]"
      },
      "all": 2799
    }
```


edit: 
**When `reads` occurs the reducer state stores `1505` instances**

edit2:
```js
const RERUN_SELECTORS = 100000
const rerunSelectors = callback => Array.from({ length: RERUN_SELECTORS }).reduce(callback)

const map = new Map(Array.from({ length: 1505 }).map((value, index) => [index, { index, object: `value-${index}` }] ))
const timer = Date.now()
console.log(rerunSelectors(() => map.get(292)))
console.log(`${(Date.now() - timer) / RERUN_SELECTORS}ms / op [1]`)
```
> 0.00016ms / op [1]

**Comment by fabienjuif** [comment]
With last commit.

What is your though @bpetetot @frinyvonnick @EmrysMyrddin  ?

```json
    results (ms): {
      "writes": {
        "total": 436,
        "init": "0.03ms / op [1000]",
        "sets": "0.11ms / op [1000]",
        "add": "0.24ms / op [1]",
        "adds": "0.48ms / op [1000]",
        "update": "0.41ms / op [1]",
        "updates": "1.38ms / op [1000]",
        "remove": "0.3ms / op [1]",
        "removes": "0.25ms / op [1000]",
        "addOrUpdate": "0.15ms / op [1]",
        "addOrUpdates": "1.01ms / op [1000]"
      },
      "reads": {
        "total": 2025,
        "getState": "0.00013ms / op [1]",
        "getKeys": "0.00015ms / op [1000]",
        "getAsArray": "0.00014ms / op [1]",
        "getLength": "0.00014ms / op [1]",
        "isInitialized": "0.00013ms / op [1]",
        "get": "0.00055ms / op [1]",
        "getById": "0.00074ms / op [1]",
        "getByIds": "0.0177ms / op [1000]",
        "hasKey": "0.00057ms / op [1]"
      },
      "all": 2461
    }
```

This is master:

```json
   results (ms): {
      "writes": {
        "total": 1059,
        "init": "0.03ms / op [1000]",
        "sets": "0.11ms / op [1000]",
        "add": "0.26ms / op [1]",
        "adds": "1.42ms / op [1000]",
        "update": "1.16ms / op [1]",
        "updates": "2.61ms / op [1000]",
        "remove": "1.37ms / op [1]",
        "removes": "1.04ms / op [1000]",
        "addOrUpdate": "0.6ms / op [1]",
        "addOrUpdates": "1.99ms / op [1000]"
      },
      "reads": {
        "total": 21410,
        "getState": "0.00013ms / op [1]",
        "getKeys": "0.00014ms / op [1000]",
        "getAsArray": "0.00015ms / op [1]",
        "getLength": "0.00016ms / op [1]",
        "isInitialized": "0.00015ms / op [1]",
        "get": "0.00058ms / op [1]",
        "getById": "0.00077ms / op [1]",
        "getByIds": "0.20996ms / op [1000]",
        "hasKey": "0.00206ms / op [1]"
      },
      "all": 22469
    }
```


edit: 
**When `reads` occurs the reducer state stores `1505` instances**

edit2:
```js
const RERUN_SELECTORS = 100000
const rerunSelectors = callback => Array.from({ length: RERUN_SELECTORS }).reduce(callback)

const map = new Map(Array.from({ length: 1505 }).map((value, index) => [index, { index, object: `value-${index}` }] ))
const timer = Date.now()
console.log(rerunSelectors(() => map.get(292)))
console.log(`${(Date.now() - timer) / RERUN_SELECTORS}ms / op [1]`)
```
> 0.00016ms / op [1]

(sorry for multiple pings)

**Comment by fabienjuif** [comment]
I will not fix coverage

**Comment by fabienjuif** [comment]
Just tester with `k-ramel/todo-mvc` and it works well 👍

**Comment by frinyvonnick** [comment]
Will you do the documentation in another PR ?

**Comment by fabienjuif** [comment]
@frinyvonnick No I will do it here.
What do you think about ?

**Comment by fabienjuif** [comment]
Do you need I publish a new RC on top of this PR ?

@frinyvonnick ?

**Comment by frinyvonnick** [comment]
Do you think its necessary to write some warning about breaking change ?

**Comment by frinyvonnick** [comment]
Sure i speak about adding some section in README.md like "Migrating from .. to .."

**Comment by fabienjuif** [comment]
Yes you are right.

I will do this in an other PR then.
So this one could be merged and tested in a new RC.

**Comment by fabienjuif** [comment]
@frinyvonnick do you want to review this PR ?

**Pull request merged**